File: clients/python/signalboxclient/product_recommenders.py
from typing import List
from .client_functions import get, post, delete
from .exceptions import SignalBoxException
from .target_variables import create_recommender_target_variable_value, get_recommender_target_variable_values


def create_product_recommender(access_token: str, base_url: str, common_id: str, name: str, product_ids: List[str] = None):
    json_params = {
        "commonId": common_id,
        "name": name,
        "productIds": product_ids
    }
    r = post(f'{base_url}/api/recommenders/ProductRecommenders/',
             json_params, access_token)
    if r.ok:
        return r.json()
    else:
        raise SignalBoxException(r.text)


def get_product_recommender(access_token: str, base_url: str, id):
    r = get(f'{base_url}/api/recommenders/ProductRecommenders/{id}', access_token)
    if r.ok:
        return r.json()
    else:
        raise SignalBoxException(r.text)


def invoke_product_recommender(access_token: str, base_url: str, id: int, common_user_id: str):
    json_params = {
        "commonUserId": common_user_id
    }
    r = post(f'{base_url}/api/recommenders/ProductRecommenders/{id}/invoke',
             json_params, access_token)
    if r.ok:
        return r.json()
    else:
        raise SignalBoxException(r.text)


def get_parameterset_recommender(access_token: str, base_url: str, id):
    r = delete(
        f'{base_url}/api/recommenders/ProductRecommenders/{id}', access_token)
    if r.ok:
        return r.json()
    else:
        raise SignalBoxException(r.text)


def delete_product_recommender(access_token: str, base_url: str, id):
    r = delete(
        f'{base_url}/api/recommenders/ProductRecommenders/{id}', access_token)
    if r.ok:
        return r.json()
    else:
        raise SignalBoxException(r.text)


def create_target_variable_value(access_token: str, base_url: str, id: int, start, end, name: str, value: float):
    return create_recommender_target_variable_value(access_token, base_url, "ProductRecommenders", id, start, end, name, value)


def get_target_variable_values(access_token: str, base_url: str, id: int,  name: str = None):
    return get_recommender_target_variable_values(access_token, base_url, "ProductRecommenders", id, name)

File: clients/js/src/src/api/recommenders/productRecommenders.js
import { pageQuery } from "../paging";
import { getUrl } from "../../baseUrl";
import { headers } from "../headers";

import {
  fetchLinkedRegisteredModelAsync,
  createLinkedRegisteredModelAsync,
} from "./common/linkRegisteredModels";

import * as tv from "./common/targetvariables";
import * as il from "./common/invokationLogs";

export const fetchProductRecommenders = async ({
  success,
  error,
  token,
  page,
}) => {
  const url = getUrl("api/recommenders/ProductRecommenders");
  const response = await fetch(`${url}?${pageQuery(page)}`, {
    headers: headers(token),
  });
  if (response.ok) {
    success(await response.json());
  } else {
    error(await response.json());
  }
};

export const fetchProductRecommender = async ({
  success,
  error,
  token,
  id,
}) => {
  const url = getUrl(`api/recommenders/ProductRecommenders/${id}`);
  const response = await fetch(url, {
    headers: headers(token),
  });
  if (response.ok) {
    success(await response.json());
  } else {
    error(await response.json());
  }
};

export const fetchProductRecommendations = async ({
  success,
  error,
  token,
  page,
  id,
}) => {
  const url = getUrl(
    `api/recommenders/ProductRecommenders/${id}/recommendations`
  );
  const response = await fetch(`${url}?${pageQuery(page)}`, {
    headers: headers(token),
  });
  if (response.ok) {
    success(await response.json());
  } else {
    error(await response.json());
  }
};

export const deleteProductRecommender = async ({
  success,
  error,
  token,
  id,
}) => {
  const url = getUrl(`api/recommenders/ProductRecommenders/${id}`);
  const response = await fetch(url, {
    headers: headers(token),
    method: "delete",
  });
  if (response.ok) {
    success(await response.json());
  } else {
    error(await response.json());
  }
};

export const createProductRecommender = async ({
  success,
  error,
  token,
  payload,
  onFinally,
}) => {
  try {
    const url = getUrl("api/recommenders/ProductRecommenders");
    const response = await fetch(url, {
      headers: headers(token),
      method: "post",
      body: JSON.stringify(payload),
    });
    if (response.ok) {
      success(await response.json());
    } else {
      error(await response.json());
    }
  } finally {
    if (onFinally) {
      onFinally();
    }
  }
};

export const createLinkRegisteredModel = async ({
  success,
  error,
  token,
  id,
  modelId,
}) => {
  createLinkedRegisteredModelAsync({
    recommenderApiName: "ProductRecommenders",
    id,
    modelId,
    token,
  })
    .then(success)
    .then(error);
};

export const fetchLinkedRegisteredModel = async ({
  success,
  error,
  token,
  id,
}) => {
  fetchLinkedRegisteredModelAsync({
    recommenderApiName: "ProductRecommenders",
    id,
    token,
  })
    .then(success)
    .catch(error);
};

export const invokeProductRecommender = async ({
  success,
  error,
  onFinally,
  token,
  id,
  version,
  input,
}) => {
  try {
    const url = getUrl(`api/recommenders/ProductRecommenders/${id}/invoke`);
    const result = await fetch(`${url}?version=${version || "default"}`, {
      headers: headers(token),
      method: "post",
      body: JSON.stringify(input),
    });
    if (result.ok) {
      success(await result.json());
    } else {
      error(await result.json());
    }
  } finally {
    if (onFinally) {
      onFinally();
    }
  }
};

export const fetchInvokationLogsAsync = async ({ id, token, page }) => {
  return await il.fetchRecommenderInvokationLogsAsync({
    recommenderApiName: "ProductRecommenders",
    id,
    token,
    page,
  });
};

export const fetchTargetVariablesAsync = async ({ id, token, name }) => {
  return await tv.fetchRecommenderTargetVariableValuesAsync({
    recommenderApiName: "ProductRecommenders",
    id,
    token,
    name,
  });
};

export const createTargetVariableAsync = async ({
  id,
  token,
  targetVariableValue,
}) => {
  return await tv.createRecommenderTargetVariableValueAsync({
    recommenderApiName: "ProductRecommenders",
    id,
    token,
    targetVariableValue,
  });
};
